### ISSUE TITLE
All the subscripted records from Context Broker is not in Quantum Leap when on high loads

### ISSUE THREAD

**Issue by ccsr** [orchestracities/ngsi-timeseries-api#349]
**Describe the bug**
Not all the records are found in Quantum Leap when subscribed to Context Broker on high burst load. Some records are missing from Quantum Leap

**To Reproduce**
Steps to reproduce the behaviour:
Set up simple set up with "Context Broker->Quantum Leap (CrateDB)" 
Create a subscription for a tenant from Context Broker to Quantum Leap
Start sending data by using JMeter or attached scripts with 500 concurrency. If the attached scripts were used then run in ubuntu test terminal with command "sh meta4-loop.sh" and it will also generate csv file for every concurrency. 

[scripts.zip](https://github.com/smartsdk/ngsi-timeseries-api/files/5067456/scripts.zip)

**Expected behavior**
The number of entities (records) in Context Brokers is not same as Quantum Leap

**Environment (please complete the following information):**
Ubuntu
Latest OrionV2 and all the dependencies
Latest QL and all the dependencies

**Comment by chicco785** [comment]
thx @ccsr 
in your tests, what are the resources allocated to crate? be a ware that recommended size for node is 8CPU 32GB

**Comment by pinkerltm** [comment]
We experienced similar behaviour with a much smaller update rate of about 200 Entitites per second (but Crate also had not very much ressources...3 Nodes with 2CPUs and 12GB each which we are currently doubling).

For bulk importing we also made a PoC where we write directly to crate bypassing orion (and therefore QL) and utilize the "bulk_args" propterty of crates HTTP API (https://crate.io/docs/crate/reference/en/4.2/interfaces/http.html#bulk-operations) which turned out to work very well. We could easily handle the load with our low ressources by just aggregating bulks of inserts.

I am not really sure how QL would be able to use this - you would need some kind of queue/cache which gets inserted on crate periodically - nevertheless what I find problematic at this issue after all, is that it is NOT guaranteed, that an Entity change is persited even when Orion is logging the call on the notify API and returning HTTP 200 to the Entity update call.

There is just no chance to verify if something did go wrong.

**Comment by chicco785** [comment]
@pinkerltm crates does everything in memory, and it also seems that garbage collection is not very efficient.
also with a queue (which could is anyhow an interesting solution) the problem may not be solved: in fact data are not injected when a given table is not "green" this happens when all records are not fully replicated which of course it is easier to happen when a given table receives many updates. so you would need a mechanism where when you get the error due to tables not ready, you wait and retry.

also the queue will bring up a new problem, now when a notification cannot be injected in crate, the error is immediately returned to orion (or should be, if that's not the case, there is a bug). with a queue, you would make the communication asynch, which makes impossible to return the information to orion.

**Comment by pinkerltm** [comment]
@chicco785 Yes, I understand that a queue possibly wouldn't solve this problem... we just observed that crate handles a much higher load with lesser ressources when you have any chance to bulk_insert rows.

What I wanted to add further to THIS issue: We also observed not only that rows weren't inserted on high loads, but also that rows were duplicated (which can be easily worked around by setting a primary key)

**Comment by chicco785** [comment]
@pinkerltm are you using context broker in high availability, i.e. two or more instances? we had duplicates only when by chance two instances of context broker generated parallel notifications (because they both picked up the notification before tagging it)

**Comment by pinkerltm** [comment]
Yes, three instances. I thought the node which receives the Entity update is also responsible for notifying the subscribers (?) Luckily it turned out as a minor issue by just setting the primary key (no duplicates since then, and pk violation errors are just ignored ;) )

**Comment by ccsr** [comment]
> thx @ccsr
> in your tests, what are the resources allocated to crate? be a ware that recommended size for node is 8CPU 32GB

The VM has 32GB RAM and 16 core

**Comment by chicco785** [comment]
for production purposes, it's better to wrap ql in a wsgi server, as clearly indicated by flask doc.

in this branch: https://github.com/smartsdk/ngsi-timeseries-api/tree/add-loadl-test
you can find a dockerfile version for the purpose. the code was also a bit tweaked to remove redundant select during insert (not tested in depth though - there may be issues if two concurrent updates stating a different type for the same attribute pop up, so be careful!) shortly, there should be an image with tag load-test including the changes.

on my local machine i cat get close to 100 req/s without any data loss:
```
    ✓ status was 200

    checks.....................: 100.00% ✓ 3879  ✗ 0    
    data_received..............: 758 kB  19 kB/s
    data_sent..................: 1.2 MB  29 kB/s
    http_req_blocked...........: avg=11.45ms min=0s           med=6.17µs   max=1.03s    p(90)=12.77ms  p(95)=105.43ms
    http_req_connecting........: avg=10.59ms min=-15.088991ms med=0s       max=1.03s    p(90)=9.57ms   p(95)=101.43ms
    http_req_duration..........: avg=4.46s   min=231.62ms     med=2.94s    max=19.28s   p(90)=10.29s   p(95)=13.27s  
    http_req_receiving.........: avg=3.78ms  min=23.6µs       med=169.62µs max=193.17ms p(90)=10.21ms  p(95)=19.91ms 
    http_req_sending...........: avg=4.48ms  min=10.11µs      med=39.84µs  max=542.48ms p(90)=650.07µs p(95)=9.61ms  
    http_req_tls_handshaking...: avg=0s      min=0s           med=0s       max=0s       p(90)=0s       p(95)=0s      
    http_req_waiting...........: avg=4.45s   min=229.32ms     med=2.94s    max=19.28s   p(90)=10.28s   p(95)=13.23s  
    http_reqs..................: 3879    95.624974/s
    vus........................: 500     min=500 max=500
    vus_max....................: 500     min=500 max=500
```

clearly `http_req_waiting` tell us that there is along queue on the system. in my case response time degrades with the load because crate insert query response time degrade (locally i gave only 1 GB heap), in some point to 1sec per insert

![Screenshot 2020-08-15 at 04 15 27](https://user-images.githubusercontent.com/777218/90303413-11a62b80-deae-11ea-896b-0b0df31c609d.png)


@ccsr i couldn't use on mac your script. so i did something similar using k6.io.

clearly you can optmize gunicorn configuration. the one included in the docker:
```
gunicorn -b 0.0.0.0:8668 uwsgi --log-level DEBUG --worker-class gevent --worker-connections 10000 --config gunicorn.conf.py
```

gunicorn.conf.py:
```
import multiprocessing

workers = multiprocessing.cpu_count() * 4 + 1
```

@pinkerltm ql already support crate native batch insert, the issue is that it works only when orion sends batch inputs

**Comment by profijoeln** [comment]
Hello, we have tried this load-test image (`smartsdk/quantumleap:load-test@sha256:edc18cfbcd77f7e0d1053e1db1d18fc5e48164303cbe2598b26883f4ce0562ef`) but with this image Orion CB receives a respond code 500 from quantumleap:

`time=Thursday 24 Sep 05:59:39 2020.408Z | lvl=WARN | corr=21db6118-fe2b-11ea-a73c-0242ac280018 | trans=1600858489-368-00000459742 | from=10.0.0.2 | srv=testqlnew | subsrv=<none> | comp=Orion | op=httpRequestSend.cpp[583]:httpRequestSendWithCurl | msg=Notification response NOT OK, http code: 500`

...and data is not present in CrateDB.

Does this build need some additional configuration changes to work ? @chicco785

**Comment by chicco785** [comment]
@profijoeln we merged on master the wsgi support which should help with performance (assuming crate has enough resources to copy with that - in our tests performance bottle necks boils down to crate), the image you reference was very experimental and had some tweaks to cache in memory configuration queries, which is not ideal since it may generate concurrency issues. we still think that caching is an interesting solution to reduce load on crate and speed up inserts, but it requires a proper solution. eventually, given redis is already integrated to cache geocoding requests, we could exploit that.

bottomline, the branch as it this will not be integrated, it served as only as experiment to evaluate how / what can be done to increase insert load on quantumleap side. at the time being, I don't recall any trick needed to make it work, but it is also true that it was tested only against the load test, which may means that the caching tweaks may have broken other parts.

**Comment by chicco785** [comment]
@profijoeln please have a look at PR #373 that should provide a more solid caching based on redis. i triggered also the public build of an image based on it with tag `smartsdk/quantumleap:redis-cache` to test it properly you would need redis, but beyond that, no specific configuration is needed.

despite that introducing redis generates of course some latency, thanks to other optimization on connection management, the throughput is the same as all in memory:

```
docker run -i --rm loadimpact/k6 run --vus 30 --duration 60s - < notify-load-test.js

    checks.....................: 100.00% ✓ 6000 ✗ 0   
    data_received..............: 1.6 MB  27 kB/s
    data_sent..................: 1.6 MB  27 kB/s
    http_req_blocked...........: avg=302.26µs min=1.62µs      med=6.21µs   max=1.05s   p(90)=10.21µs  p(95)=18µs    
    http_req_connecting........: avg=262.78µs min=0s          med=0s       max=1.05s   p(90)=0s       p(95)=0s      
    http_req_duration..........: avg=208.12ms min=16.92ms     med=155.56ms max=1.06s   p(90)=409.38ms p(95)=514.53ms
    http_req_receiving.........: avg=1.2ms    min=-8.817564ms med=158.96µs max=133.5ms p(90)=2.69ms   p(95)=5.99ms  
    http_req_sending...........: avg=89.36µs  min=12.59µs     med=38.08µs  max=20.85ms p(90)=111.97µs p(95)=218.72µs
    http_req_tls_handshaking...: avg=0s       min=0s          med=0s       max=0s      p(90)=0s       p(95)=0s      
    http_req_waiting...........: avg=206.82ms min=16.78ms     med=153.98ms max=1.06s   p(90)=406.54ms p(95)=513.29ms
    http_reqs..................: 6000    99.814509/s
    iteration_duration.........: avg=30.02s   min=30.02s      med=30.02s   max=30.03s  p(90)=30.02s   p(95)=30.02s  
    iterations.................: 60      0.998145/s
    vus........................: 30      min=30 max=30
    vus_max....................: 30      min=30 max=30
```

**Comment by profijoeln** [comment]
Hey @chicco785 !

Does your loadimpact test run for 60 seconds? Or do I misunderstand how this test works.

In our testing we have not limited CrateDB resources (or any component) and we have no indication of high resource use to begin with.  We have witnessed data-loss in longer tests like 24 hour tests of sending real data at speeds 60 msg/s. In OrionCB we check subscriptions to see that Orion has notified of each entry, but all those entries cannot be found in CrateDB. We have witnessed around 30 messages lost in 5 million messages and sometimes few in just shorter 1 hour testing. We have not witnessed an error in QuantumLeap or CrateDB logs of something going wrong. In our tests we have also monitored CPU and RAM usage to make sure components are not running out of resources. So when using architecture where Context Broker notifies Quantumleap with entities, even with more than enough resources available, we cannot guarantee message delivery and persistence.

When we use higher speeds of what we see as the limit of Quantumleap, we see some sort of queue in QuantumLeap, in form of PIDs and increasing RAM usage. If we use too high throughput processes pile up and Quantumleap appears to be handling them as fast as it can, but not enough to create alarming load on CrateDB. Once data stops flowing in Quantumleap is still processing those queued up messages, this is what we call post-processing. Load from QuantumLeap to CrateDB continues as long as there are PIDs in the queue. Only once this entry is fully processed QuantumLeap responds to Orion, usually at this point resulting into timeout, since they might have been in the queue for a long time.

**Comment by chicco785** [comment]
@profijoeln 

the test runs for 60 seconds, plus 30 seconds of cool down.

vus are single clients
each client tries to send 100 updates in a row, as you can see in  `notify-load-test.js`

quantumleap has no queue itself, but https://gunicorn.org/ - the wsgi server has. so, till the ql process is not free it keeps up to 100 requests (or something like that), you should check in the gunicorn defaults. of course you can also change gunicorn configuration if needed (customising this file: https://github.com/smartsdk/ngsi-timeseries-api/blob/master/src/server/gconfig.py).

Understanding why and how 30 messages got lost over 5 million, is not that easy. Where you using debug configuration in QL?

In general, 100% reliable messaging with the way orion provides notification is not possible, you would need something different from http, with possibility of ensuring message transport (e.g. amqp or mqtt), i.e. you would need a different architecture. 

Without details and understanding of your deployment, it is hard to help. 
Few things we noticed:
- under heavy load crate health check return 500 (or nothing), thus causing k8s health check  probes on quantumleap to fail. of course, when this happens, messages don't reach quantumleap.
- orion with ha deployment during heavy load may cause weird behaviours: a message notified twice because two orion nodes processed the message in parallel before the message was marked as notified, or a message not notified, because before notification was triggered a new update for the same entity came in.

That's why the better way to understand if there are issues on ql side, is testing notification directly on it, not through orion which may bring up other type of issues.

To run a better performance assessment, we would need to meter time to process for each request - something that should not be too complex to add. It could be that part of the overhead is due to message validation by flask. switching to FastAPI may help, but we haven't yet evaluated it.

**Comment by chicco785** [comment]
@profijoeln 
running 24h test on my laptop is probably out of question as you can imagine. still i have been doing few additional tests, with longer duration:

```
docker run -i --rm loadimpact/k6 run --vus 400 --duration 600s - < notify-load-test.js

    ✓ status was 200

    checks.....................: 100.00% ✓ 66658 ✗ 0    
    data_received..............: 13 MB   21 kB/s
    data_sent..................: 18 MB   29 kB/s
    http_req_blocked...........: avg=1.29ms   min=19.5µs     med=35.5µs  max=1.04s    p(90)=77.8µs  p(95)=239.71µs
    http_req_connecting........: avg=1.13ms   min=0s         med=0s      max=938.34ms p(90)=0s      p(95)=0s      
    http_req_duration..........: avg=3.74s    min=563.31ms   med=3.27s   max=16.49s   p(90)=6.37s   p(95)=7.18s   
    http_req_receiving.........: avg=2.13ms   min=-31.3816ms med=303.9µs max=203.51ms p(90)=5.81ms  p(95)=10.64ms 
    http_req_sending...........: avg=362.61µs min=-33.617ms  med=74.7µs  max=716.13ms p(90)=309.2µs p(95)=653.01µs
    http_req_tls_handshaking...: avg=0s       min=0s         med=0s      max=0s       p(90)=0s      p(95)=0s      
    http_req_waiting...........: avg=3.74s    min=563.03ms   med=3.27s   max=16.43s   p(90)=6.37s   p(95)=7.17s   
    http_reqs..................: 66658   105.826468/s
    iteration_duration.........: avg=5m37s    min=1m57s      med=5m25s   max=8m58s    p(90)=8m45s   p(95)=8m56s   
    iterations.................: 449     0.712834/s
    vus........................: 381     min=381 max=400
    vus_max....................: 400     min=400 max=400
```

so, clearly hitting 105 req/s is not an issue on my mac (using docker with ql+redis+crate), which makes me think on a production server, you can get much more if needed.

still, as you can see what is not nice is the avg duration:
`avg=3.74s    min=563.31ms   med=3.27s   max=16.49s   p(90)=6.37s   p(95)=7.18s`

nevertheless, logging "internally" the processing takes in average 80 msec (out which 50 msec is the time for crate insert), see below a graph
![logs sx (1)](https://user-images.githubusercontent.com/777218/95261992-aaf91a80-082b-11eb-9871-76fe6375c0c9.png)

so what's adding the 3 seconds? a profiler would be of help to get deeper, but gunicorn + flask may have quite a role in that, as well as python garbage collection. surely, fine tuning gunicorn configuration may be quite important to achieve best performance.

as said, for performance, flask seems to be not ideal and a good alternative could be https://fastapi.tiangolo.com/
still,  105 reqs per second on a laptop does not look bad. at the time being, we will work on consolidating pr #373, but we won't invest on fastapi (we don't have a the moment a use case / customer that justify that investment).

**Comment by c0c0n3** [comment]
> so what's adding the 3 seconds? a profiler would be of help to get deeper

So I've done some preliminary profiling, see https://github.com/smartsdk/ngsi-timeseries-api/issues/399#issuecomment-733708372

**Comment by chicco785** [comment]
> I am not really sure how QL would be able to use this - you would need some kind of queue/cache which gets inserted on crate periodically - nevertheless what I find problematic at this issue after all, is that it is NOT guaranteed, that an Entity change is persited even when Orion is logging the call on the notify API and returning HTTP 200 to the Entity update call.

@pinkerltm for your information, @ilarimikkonen sponsored a queue-based version of ql which is available since 0.8.2, you can refer to the docs: 
- https://quantumleap.readthedocs.io/en/latest/admin/installing/#deploy-quantumleap-with-the-work-queue
- https://quantumleap.readthedocs.io/en/latest/admin/wq/
